### ISSUE TITLE
Externals of type this don't play well with module concatenation

### ISSUE THREAD

**Issue by jsnajdr** [webpack/webpack#11724]
# Bug report

**What is the current behavior?**

Compile this source _src/index.js_:
```js
import { map } from 'lodash';
map( [], a => a + 1 );
```

with this _webpack.config.js_:
```js
module.exports = {
  optimization: {
    minimize: false
  },
  externals: {
    lodash: 'this lodash'
  }
}
```

The `lodash` import should point to `this.lodash` -- at runtime, lodash is already loaded with a `<script>` tag that assigns `lodash` to the global `this`, i.e., either `window` or `global`.

The resulting _dist/main.js_ however looks like this:
```js
/******/ (() => { // webpackBootstrap
/******/ 	"use strict";

// CONCATENATED MODULE: external "lodash"
(function() { const __WEBPACK_NAMESPACE_OBJECT__ = this["lodash"]; })();
// CONCATENATED MODULE: ./src/index.js

(0,undefined.map)( [], a => a + 1 );

/******/ })()
;
```

There are two issues here:
1. the import and the usage are not connected together: it's `__WEBPACK_NAMESPACE_OBJECT__` vs `undefined`. If I use external of type `window`, both are connected correctly with a `external_lodash_namespaceObject` variable. The only difference between `window` and `this` type seems to be that `this` is wrapped in an IIFE. That apparently breaks things.
2. the bundle is in strict mode (because the `allStrict` condition holds, i.e., all modules have `m.buildInfo.strict === true`)
 and `this` is therefore `undefined`, not the global this. That breaks any attempt to use `externalType: 'this'`. Maybe `ExternalModule` of type `this` should have `buildInfo.strict === false`?

Disabling `concatenateModules: false` fixes issue no.1, but issue no.2 is still there.

In webpack 4 this used to work because:
- there was no bundle-global `use strict`
- external modules bailed out of module concatenation with the "Module is not an ECMAScript module" reason

**If the current behavior is a bug, please provide the steps to reproduce.**
A repo with the above example: https://github.com/jsnajdr/webpack-externals-concat-bug

Just `npm install` and then run webpack.

**Other relevant information:**
webpack version:  5.1.3

**Comment by code1x1** [comment]
i forked your project and did minor changes to make it a working example. even with html template and external lodash script import

https://github.com/code1x1/webpack-externals-concat-bug/tree/master

you assume that the `undefined.map` doesnt map to the `this['lodash']` module but it did and webpack did assume that `this['lodash']` will be `undefined` because this is scoped to the module closure. the webpack bundler know that `this` will not contain a lodash property in strict mode hence it will evaluate it to undefined.
the problem is that your setup is not possilbe. the main reason is when you would drop the strict mode all your modules will be in one giant this object. correct me if im missing some logic here.

**Comment by jsnajdr** [comment]
@code1x1 While it's true that various parts of webpack evaluate JS expressions and can tell that something is always `undefined`, I don't think it's happening here. The bug is that the variable that the exports are assigned to is inside a function (the IIFE) and therefore not visible outside the function. That breaks the concatenation logic.

> the main reason is when you would drop the strict mode all your modules will be in one giant this object.

If I drop the strict mode just for one module, it means that webpack will need to add `"use strict"` at the beginning of every module function, instead of adding just one global one at the top.

That's really happening and I'd say it's the cost of using `this` as a global. It works only in non-strict mode and doesn't make sense in strict mode.

We'll probably move away from using the `this` global in our project, as it inevitably de-optimizes the bundle, but `external: this` continues to be a supported feature of webpack and it should work.

I submitted a fix in #11735 and described more implementation details there.
